File: LeetCode/Problems/0026_Remove_Duplicates_from_Sorted_Array/remove_duplicates_from_sorted_array.py
from typing import List

class Solution:
    # time complexity: O(n), space complexity: O(1)
    def removeDuplicates(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0

        i = 0
        for n in nums:
            if(nums[i] != n):
                i += 1
                nums[i] = n

        return i+1

def printList(nums: List[int]) -> str:
    return('[' + ', '.join(map(str, nums)) + ']')

nums = [1,1,2]
print("Input: " + printList(nums))
size = Solution().removeDuplicates(nums)
print("Output: " + printList(nums[:size]))

nums = [0,0,1,1,1,2,2,3,3,4]
print("\nInput: " + printList(nums))
size = Solution().removeDuplicates(nums)
print("Output: " + printList(nums[:size]))

File: LeetCode/Problems/0026_Remove_Duplicates_from_Sorted_Array/remove_duplicates_from_sorted_array.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        return twoPointers(nums);
    }

private:
    // time complexity: O(n), space complexity: O(1)
    int twoPointers(vector<int>& nums){
        if(nums.empty())
            return 0;

        int i = 0;
        for(int j = 1; j < nums.size(); j++)
            if(nums[i] != nums[j])
                nums[++i] = nums[j];

        return i+1;
    }

    int twoPointers2(vector<int>& nums){
        if(nums.empty())
            return 0;

        int i = 0;
        for(const auto& n : nums)
            if(nums[i] != n)
                nums[++i] = n;

        return i+1;
    }
};

// nums is passed by constant reference and cannot be changed by this function
string printVector(const vector<int>& nums, int len){
    string str;

    str.append("[");
    for(int i = 0; i < len; i++){
        if(i != 0)
            str.append(", ");
        str.append(std::to_string(nums[i]));
    }
    str.append("]");

    return str;
}

int main(){
    vector<int> nums({1, 1, 2});
    cout << "Input: " << printVector(nums, 3) << "\n";

    Solution sol;
    int len = sol.removeDuplicates(nums);
    cout << "Output: " << printVector(nums, len) << "\n";

    return 0;
}
